### PR TITLE
US2123827: abort in flight coroutine

### PR DESCRIPTION
**Pull request by abaeza-wp**
### What
- Reviewed all coroutines and paths to ensure consistent behaviours, and ensure co-routines are launched and switch the the correct dispatchers
- Use Dependency injection: singletons to be able to provide configuration while service discovery is not available and to be able to override the dispatchers with test dispatchers
- Adding a `SupervisorJob` to the `CoroutineScope` in the `CardBinService` class. This ensures that child coroutines can fail independently without cancelling the entire scope.
- Addressed an edge case where `globalBrand` could be null and enhanced card-bin logic to handle unmatched response brands.
- Implemented cache manager for card-bin-service responses
- Refactored the mock servers into their own class to isolate bahaviours and have a cleaner base class
- Refactored coroutine tests to use BaseCoroutineTest()
- Fixed a race condition in `CardConfigurationProvider` that caused incorrect brand responses to be cached.
- Improved unit tests for stability and added cases for null `globalBrand`.
- Fixed an issue in the demo app where images were not updated on the UI thread and improved rendering logic.
- Updated tests to use `runTest` instead of `runBlocking`.
- Updated and reviewed all tests to ensure stability

### Why:
- To ensure the application handles edge cases gracefully and avoids incorrect caching of brand responses.
- To improve test reliability and coverage for new behaviours.
- To fix UI issues in the demo app.
- To improve test maintainability.
- Currently, if a child coroutine fails or is cancelled, it cancels the entire `CoroutineScope`. This prevents launching new coroutines, which prevents the field from making requests to CardBinService. 

### How:
- Updated `CardConfigurationProvider` to use `runBlocking` with an IO-scoped coroutine to prevent race conditions.
- Enhanced card-bin logic to transform response brands and handle unmatched cases.
- Refactored and added unit tests to cover null `globalBrand` scenarios and improve stability.
- Fixed demo app image updates by ensuring operations run on the UI thread and improved rendering logic for unknown card icons and labels.
- Moved `BaseCoroutineTest` to a shared location and updated tests to use `runTest` for better coroutine handling.
- Changed `scope` initialization to include a `SupervisorJob`. This ensures that the parent scope is not cancelled when a child coroutine fails.

### CODE REVIEW

File: demo-app/src/androidTestMock/java/com/worldpay/access/checkout/sample/cvc/CvcFragmentTestUtils.kt [hunk 1]
```diff
@@ -66,7 +66,11 @@ class CvcFragmentTestUtils(activityRule: ActivityTestRule<MainActivity>) : Abstr
     }
 
     fun focusOff(): CvcFragmentTestUtils {
-        cvcInput().onFocusChangeListener?.onFocusChange(cvcInput(), false)
+        // Naugthy code:
```
Comment (jmensahWP) [inline demo-app/src/androidTestMock/java/com/worldpay/access/checkout/sample/cvc/CvcFragmentTestUtils.kt:69]: Naughty code lol

File: access-checkout/src/test/java/com/worldpay/access/checkout/cardbin/service/CardBinServiceTest.kt [hunk 2]
```diff
@@ -300,44 +183,225 @@ class CardBinServiceTest : BaseCoroutineTest() {
...
+            val mockClient = mock<CardBinClient>()
+            val mockJob = mock<Job>()
+            val cardBinService = CardBinService(
+                checkoutId = checkoutId,
+                baseUrl = baseUrl,
+                client = mockClient
+            )
+            cardBinService.currentJob = mockJob // Set the current job to the mocked job
+
+            assertEquals(mockJob, cardBinService.currentJob)
```
Comment (abaeza-wp) [inline access-checkout/src/test/java/com/worldpay/access/checkout/cardbin/service/CardBinServiceTest.kt:405]: double check if we can remove now as not neccessary

File: access-checkout/src/test/java/com/worldpay/access/checkout/client/testutil/AbstractBaseCoroutineTest.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            Dispatchers.setMain(testDispatcher)
+        }
+
+        fun coroutineTearDown() {
+            testScope.cancel()
+            Dispatchers.resetMain()
+            DispatchersProvider.instance = CoroutineDispatchers()
+        }
+    }
+}
```
Comment (abaeza-wp) [inline access-checkout/src/test/java/com/worldpay/access/checkout/client/testutil/AbstractBaseCoroutineTest.kt:47]: can we. re-use BaseCoroutineTest

File: demo-app/src/androidTestMock/java/com/worldpay/access/checkout/sample/cvc/AbstractCvcFragmentTest.kt [hunk 4]
```diff
@@ -24,8 +24,8 @@ abstract class AbstractCvcFragmentTest {
         cvcFragmentTestUtils = CvcFragmentTestUtils(activityRule)
         defaultStubMappings(activityRule.activity)
         navigateTo(R.id.nav_cvc_flow)
-        closeSoftKeyboard()
-        rotatePortrait(activityRule)
+//        closeSoftKeyboard()
+//        rotatePortrait(activityRule)
```
Comment (abaeza-wp) [inline demo-app/src/androidTestMock/java/com/worldpay/access/checkout/sample/cvc/AbstractCvcFragmentTest.kt:28]: Add comment to explain why these have been left commentted out

File: access-checkout/src/test/java/com/worldpay/access/checkout/cardbin/service/CardBinServiceTest.kt (hunk 2 above)
Comment (abaeza-wp) [reply access-checkout/src/test/java/com/worldpay/access/checkout/cardbin/service/CardBinServiceTest.kt:405]: confirmed needed for code coverage as property has been made internal just for testing purposes

File: demo-app/src/androidTestMock/java/com/worldpay/access/checkout/sample/card/standard/testutil/AbstractCardFragmentTest.kt [hunk 5]
```diff
@@ -21,8 +19,8 @@ abstract class AbstractCardFragmentTest {
     fun setup() {
         cardFragmentTestUtils = CardFragmentTestUtils(activityRule)
         defaultStubMappings(activityRule.activity)
-        closeSoftKeyboard()
-        rotatePortrait(activityRule)
+//        closeSoftKeyboard()
```
[the remark this answers is not in the document]
Comment (abaeza-wp) [reply demo-app/src/androidTestMock/java/com/worldpay/access/checkout/sample/card/standard/testutil/AbstractCardFragmentTest.kt:22]: Will add a comment:

When starting a new tests the below introduce unnecessary delays therefore I have left them commented out for now in case we ever need to re-enable, But it would be better to test this behaviour only when necessary on a per test-case basis We will review them in the future again
